crates/block-producer/src/mempool/transaction_graph.rs

```diff
@@ -0,0 +1,123 @@
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+};
+
+use miden_objects::transaction::{ProvenTransaction, TransactionId};
+
+use super::BatchJobId;
+
+#[derive(Default, Clone, Debug)]
+pub struct TransactionGraph {
+    /// All transactions currently inflight.
+    nodes: BTreeMap<TransactionId, Node>,
+
+    /// Transactions ready for inclusion in a batch.
+    ///
+    /// aka transactions whose parent transactions are already included in batches.
+    roots: BTreeSet<TransactionId>,
+}
+impl TransactionGraph {
+    pub fn insert(&mut self, transaction: ProvenTransaction, parents: BTreeSet<TransactionId>) {
+        let id = transaction.id();
+
+        // Inform parent's of their new child.
+        for parent in &parents {
+            self.nodes.get_mut(&parent).expect("Parent must be in pool").children.insert(id);
+        }
+
+        let transaction = Node {
+            status: Status::InQueue,
+            data: Arc::new(transaction),
+            parents,
+            children: Default::default(),
+        };
+        if self.nodes.insert(id, transaction).is_some() {
+            panic!("Transaction already exists in pool");
+        }
+
+        // This could be optimised by inlining this inside the parent loop. This would prevent the
+        // double iteration over parents, at the cost of some code duplication.
+        self.try_make_root(id);
+    }
+
+    pub fn pop_for_batching(
+        &mut self,
+    ) -> Option<(Arc<ProvenTransaction>, BTreeSet<TransactionId>)> {
+        let transaction = self.roots.pop_first()?;
+        let transaction =
+            self.nodes.get_mut(&transaction).expect("Root transaction must be in graph");
+        transaction.status = Status::Processed;
+
+        // Work around multiple mutable borrows of self.
+        let data = Arc::clone(&transaction.data);
+        let parents = transaction.parents.clone();
+        let children = transaction.children.clone();
+
+        for child in children {
+            self.try_make_root(child);
+        }
+
+        Some((data, parents))
+    }
+
+    /// Marks the given transactions as being back inqueue.
+    pub fn requeue_transactions(&mut self, transactions: BTreeSet<TransactionId>) {
+        for tx in &transactions {
+            self.nodes.get_mut(&tx).expect("Node must exist").status = Status::InQueue;
+        }
+
+        // All requeued transactions are potential roots, and current roots may have been
+        // invalidated.
+        let mut potential_roots = transactions;
+        potential_roots.extend(&self.roots);
+        self.roots.clear();
+        for tx in potential_roots {
+            self.try_make_root(tx);
+        }
+    }
+
+    pub fn remove_stale(&mut self, transactions: Vec<TransactionId>) {
```

Comment:
The transaction data could be removed as soon as the block is finalized. This does mean we need to make changes when new transactions enter as the "parent" transaction might no longer exist.

This is fine though it does remove an opportunity for verifying the graphs correctness at runtime since we can no longer distinguish between a broken graph and parent data just being removed. The benefit is trimming the memory usage earlier.

For this initial implementation my thought process was to ensure that the graph was always correct i.e. all indices/edges always fully exist until the entire block is removed once its stale.

Reply:
> The transaction data could be removed as soon as the block is finalized. This does mean we need to make changes when new transactions enter as the "parent" transaction might no longer exist.

Makes sense - though I was thinking we'd make this check before adding the transaction to the graph (i.e., we call `add_transaction()` only with a valid set of parents).

Reply:
I'm not sure if we're talking past each other here; so to clarify:

Inflight state contains the state of all in-progress transactions as well as some number of recently committed blocks. The latter is required so that new transactions have some grace/overlap period to pre-fetch input data from the store.

State also contains it's source transaction ID so that new transactions can be added as children to form the dependency graph.

State from stale blocks must be pruned otherwise the in-memory state will grow indefinitely and essentially be a copy of all the state ever. So if block `N` becomes "stale" (the wording here is bad since its referring to the transaction inputs being considered stale but... 🤷), all state still attached to block `N` should be removed.

Transactions in committed blocks can be deleted immediately, however this leaves the state transaction ID dangling. Here are the options to handle this:

1. Only delete transaction data once the block is stale (current approach). New transactions can depend on transactions already in committed blocks which is a bit of a waste ito compute/lookups, but shouldn't be much. Highest memory usage since txn data hangs around the longest. But graph edges remain consistent - all edges will always be valid.
2. Delete data as soon as it is committed to store. This makes invalid tx parent edges possible everywhere, meaning we cannot use `tx_graph.get(parent).expect(..)` to assert graph conditions at runtime. Instead we must allow these to fail as allowed.
3. Delete data as soon as it is committed to store. Remove edges everywhere, except state. Only the state edges are now allowed to fail. The reason we do the latter is because the state data must remain active until the stale block is removed.
    1. We can amend this by having the state contain an optional txn ID which can be removed once the block is committed. However this means each block will touch the state twice. Once when it is committed, and again when it is stale.

Maybe as an analogy, what we would be doing is changing from multiple `Rc<Transaction>` to a single `Rc` with parent and state pointers becoming `rc::Weak` pointers. Which is fine; but we do lose the ability to sanity check at runtime with possible memory leaks, and edges that should exist but no longer do.

Reply:
> (i.e., we call add_transaction() only with a valid set of parents).

I don't quite follow this point

Reply:
I was thinking of something along the lines of the 3rd option: once a block is committed we prune all transactions of this block from `tx_graph` and all batches of this block from `batch_graph`. We do keep the information about the block's created notes/nullifiers and updated accounts separately (i think this is what you refer to as `state`). These are removed at a later point in time when the block becomes "stale".

But I haven't thought about it as much as you, so - may be missing some details about this approach.

> > (i.e., we call add_transaction() only with a valid set of parents).
> 
> I don't quite follow this point

The idea here was that when we call `tx_graph.add_transaction()` the parents will contain only transactions which are currently in `tx_graph`.

Reply:
> The idea here was that when we call tx_graph.add_transaction() the parents will contain only transactions which are currently in tx_graph.

Got it; yes we can do that.

> I was thinking of something along the lines of the 3rd option: once a block is committed we prune all transactions of this block from tx_graph and all batches of this block from batch_graph. We do keep the information about the block's created notes/nullifiers and updated accounts separately (i think this is what you refer to as state). These are removed at a later point in time when the block becomes "stale".
>
> But I haven't thought about it as much as you, so - may be missing some details about this approach.

I think superficially I considered the transaction-only approach as simpler, as it doesn't have to also handle blocks. 

As a mockup:

```rust
/// Represents state tracking for a single account using tx only.
struct AccountState {
    /// Transactions and states from all inflight txns and non-stale blocks.
    /// Required to allow reverting transactions.
    states: VecDequeu<(Digest, TransactionId)>,
}

impl AccountState {
    /// Rewinds the state of all the given transactions i.e. pop_back.
    /// Could also simplify and just make it a # of txns to revert which means
    /// less checking required.
    fn revert(&mut self, txs: BTreeSet<TransactionId>) { ... }

    /// Removes the states from the from the front of the queue i.e. pop_front.
    /// Could also simplify to # here.
    fn remove_stale(&mut self, txs: BTreeSet<TransactionId>) { ... }

    /// The current state of the account.
    fn current(&self) -> Option<&(Digest, TransactionId)> { self.states.last() }

    fn insert(&mut self, state: Digest, tx: TransactionId) { self.states.push_back((state, tx)); }
}
```

Versus with block support:

```rust
struct AccountState {
    inflight: VecDequeu<(Digest, TransactionId)>,
    /// Latest complete but not-stale block.
    block: Option<Digest>,
}
```

impl AccountState {
    fn revert(&mut self, txs: BTreeSet<TransactionId>) { ... }

    fn remove_stale(&mut self, state: Digest) { 
        // Remove self.block if its state matches the input.
    }

    fn current(&self) -> Option<&(Digest, Option<TransactionId>)> { 
        self.states.last()
            .map(..)                 // map tx -> Some
            .or(
                 self.block.map(...) // add tx -> None
            )
    }

    /// Using count prevents safety checks but does make it easier to implement.
    fn block_complete(&mut self, tx_count: usize) {
         self.block = self.states
             .drain(0..tx_count)
             .last();
    }
}
```

I was expecting the block handling to be more complex tbh. The nice thing about the latter is that edges are always valid. I'll try it out and see how it goes.